### ISSUE TITLE
resize() + onlyScaleDown() should not affect cache key if no resize happens.

### ISSUE THREAD

**Issue by NightlyNexus** [square/picasso#2072]

**Comment by JakeWharton** [comment]
Makes sense. Implications are non-trivial I think...

**Comment by NightlyNexus** [comment]
yeah, pretty sure this is tricky. the cache key is created at request time for the cache time but we'd need to edit it afterward for this.
i think it's a worthwhile optimization, though. i saw a case of it in our app.

maybe a new type of cache key.

**Comment by JakeWharton** [comment]
Changing to compute the actual key which represents the image instead of
what was requested seems worthwhile

**Comment by NightlyNexus** [comment]
on the other hand, if i call `load(url).resize(500, 500).onlyScaleDown()` and then call `load(url).resize(500, 500).onlyScaleDown()` again, i don't want to redownload the image again.
the cache key would need to match on both the original request and the result for this feature.

**Comment by JakeWharton** [comment]
Yep. We can start with something specific for this case, but it'd be nice
to solve the case of loading a thumbnail from the cache for a larger image
while the larger version decodes from disk cache. We could do this either
with a series of secondary fallback keys or with some kind of cached key
matcher.
